Move datapath verifier tests into GH actions workflow

tklauser opened the pull request.

brb commented on .github/workflows/conformance-datapath.yaml:
```diff
@@ -180,6 +182,12 @@ jobs:
           docker cp $cid:/usr/local/bin/cilium ./cilium-cli
           docker rm $cid
 
+      # For `go test -c` in the verifier step below.
```
I think this is not needed, as you compile in the LVH VM, right?

tklauser replied:
Ah nice, the LVM VM has `go` as well. Is this guaranteed to be the same version as in `GO_VERSION`? Otherwise it's one repo more for me to keep track of to bump Go versions :smile:

tklauser replied:
nvm, I've misread the test step and was assuming the test to be compiled in the GH action rather than the LVH VM. I see that golang 1.19 is [used in the images](https://github.com/cilium/little-vm-helper-images/blob/main/_data/images.json), so should be fine.

brb reviewed (commented):
Sweet! Will ACK once it's ready for the review :rocket:

pchaigno commented on .github/workflows/conformance-datapath.yaml:
```diff
@@ -254,6 +256,22 @@ jobs:
+              -v /sys/fs/bpf:/sys/fs/bpf \
+              -v "\$PWD:/cilium" \
+              quay.io/cilium/test-verifier:2ecf56b4ea57576e9d92d34407898e5d14e80aa3@sha256:62396cedb4f15477f0084d7dfc92de55ac9ab8531021b7ac5f56220c35f2cb64 \
+              /cilium/verifier.test -test.v -test.parallel=1 -cilium-base-path /cilium
```
Should this be in its own workflow? The verifier tests are only going to grow over time and their scope/goal is fairly different from other datapath tests, which are end-to-end.

ti-mo replied:
Agreed, for me the term 'conformance' exists purely within the scope of k8s ([the certification program](https://www.cncf.io/certification/software-conformance/)), and I've always found it strange that the verifier tests were a part of that.

tklauser replied:
Agreed, I'll move this into its own workflow without any reference to the term "conformance".

ti-mo commented on .github/workflows/tests-datapath-verifier.yaml:
```diff
@@ -0,0 +1,237 @@
+name: Cilium Datapath BPF Verifier Complexity
```
Nit: 'Cilium' is kind of implied here. This gets quite long in the job view at the bottom and becomes too wide for the box: `Cilium Datapath BPF Verifier Complexity / Setup & Test (5.4-main) (pull_request) Successful in 4m`

```suggestion
name: Datapath BPF Complexity
```
I think this should be sufficiently specific.

pchaigno commented on test/verifier/verifier_test.go:
```diff
@@ -0,0 +1,152 @@
+	default:
+		// t.Fatalf("kernel version %s not supported by verifier complexity tests", release)
+		ciKernel = "49"
+	}
```
I'm unsure what's the best way to support cases not covered by the compilation environment variable :thinking: 

Currently, the compilation supports 4.9, 4.19, 5.4, and bpf-next. I see three options:
1. We always try to stick to the closest, inferior version. So running the test on 5.10 would compile with KERNEL=5.4.
2. We require that the compilation supports any kernel we run on. So running the test on 5.10 would currently fail with an error asking to add compilation support.
3. We pass the kernel to use as a parameter of this executable.

I would be in favor of option 2 because, in the long term, it better ensures we don't miss a complexity issue. But maybe option 1 can be done in the shorter term.

I'm also fine to handle this whole discussion as a follow up (not necessarily assigned to you oc) and fallback to 4.9 as you did here in the meantime.

tklauser replied:
I too think option 2 would be preferable. I think I mainly added the default case for local testing in the dev VM and forgot to adjust it before opening the PR. I've now pushed the following incremental diff which will make the test fail on any unsupported kernel version:

```diff
diff --git a/test/verifier/verifier_test.go b/test/verifier/verifier_test.go
index d6ae59abcc07..d43f972e104b 100644
--- a/test/verifier/verifier_test.go
+++ b/test/verifier/verifier_test.go
@@ -38,9 +38,10 @@ func getCIKernelVersion(t *testing.T) string {
                ciKernel = "419"
        case strings.HasPrefix(release, "5.4"):
                ciKernel = "54"
-       default:
-               // t.Fatalf("kernel version %s not supported by verifier complexity tests", release)
+       case strings.HasPrefix(release, "4.9"):
                ciKernel = "49"
+       default:
+               t.Fatalf("kernel version %s not supported by verifier complexity tests", release)
        }
        return ciKernel
 }
```

tklauser replied:
Turns out this approach won't work for bpf-next: https://github.com/cilium/cilium/actions/runs/3995264801/jobs/6853943129

So I'll add option 3 as well so we're able to override the kernel version in CI.

pchaigno commented on test/verifier/verifier_test.go:
```diff
@@ -0,0 +1,152 @@
+	case strings.HasPrefix(release, "5.4"):
+		ciKernel = "54"
+	default:
+		// t.Fatalf("kernel version %s not supported by verifier complexity tests", release)
```
Nit: Comment to be removed.

pchaigno commented on test/verifier/verifier_test.go:
```diff
@@ -0,0 +1,152 @@
+		//			name:      "bpf_sock",
+		//			hook:      HookCgroup,
+		//			macroName: "MAX_LB_OPTIONS",
+		//		},
```
Nit: Let's just remove.

pchaigno reviewed (approved):
LGTM assuming the last commit is removed.

tklauser committed:
test/verifier: make data path verifier complexity tests standalone

Based on K8sDatapathVerifier.

This allows to run the datapath verifier tests outside of a k8s cluster
as part of the datapath conformance workflow in GitHub actions.

Note that this test no longer covers kernel version 4.9 which
K8sDatapathVerifier did. However, support for that kernel version is
dropped in the upcoming Cilium 1.13 release.

Suggested-by: Martynas Pumputis <m@lambda.lt>
Signed-off-by: Tobias Klauser <tobias@cilium.io>

tklauser committed:
workflows: add datapath verifier test workflow

Signed-off-by: Tobias Klauser <tobias@cilium.io>

tklauser committed:
test/k8s: skip verifier test on all but 4.19

The 4.19 image used by the GHA datapath verifier workflow currently leads to
verifier issues on that version. The k8s-based tests use an older 4.19
version which does not have these issues. Thus, keep the k8s-based test
for now but restrict it to 4.19.

Signed-off-by: Tobias Klauser <tobias@cilium.io>

brb reviewed (approved):
:pizza: :partying_face:
